drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2024 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_digi_temp
+
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(stm32_digi_temp, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Constants */
+#define ONE_MHZ			1000000 /* Hz */
+#define TS1_T0_VAL0		30	/* °C */
+#define TS1_T0_VAL1		130	/* °C */
+#define SAMPLING_TIME		15	/* best precision */
+
+struct stm32_digi_temp_data {
+	struct k_sem sem_isr;
+	struct k_mutex mutex;
+
+	/* Peripheral clock frequency */
+	uint32_t pclk_freq;
+	/* Engineering value of the frequency measured at T0 in Hz */
+	uint32_t t0_freq;
+	/* Engineering value of the T0 temperature in °C */
+	uint16_t t0;
+	/* Engineering value of the ramp coefficient in Hz / °C */
+	uint16_t ramp_coeff;
+
+	/* Raw sensor value */
+	uint16_t raw;
+};
+
+struct stm32_digi_temp_config {
+	/* DTS instance. */
+	DTS_TypeDef *base;
+	/* Clock configuration. */
+	struct stm32_pclken pclken;
+};
+
+static void stm32_digi_temp_isr(const struct device *dev)
+{
+	struct stm32_digi_temp_data *data = dev->data;
+	const struct stm32_digi_temp_config *cfg = dev->config;
+	DTS_TypeDef *dts = cfg->base;
+
+	/* Clear interrupt */
+	SET_BIT(dts->ICIFR, DTS_ICIFR_TS1_CITEF);
+
+	/* Give semaphore */
+	k_sem_give(&data->sem_isr);
+}
+
+static int stm32_digi_temp_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct stm32_digi_temp_config *cfg = dev->config;
+	struct stm32_digi_temp_data *data = dev->data;
+	DTS_TypeDef *dts = cfg->base;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Wait for the sensor to be ready (~40µS delay after enabling it) */
+	while (READ_BIT(dts->SR, DTS_SR_TS1_RDY) == 0) {
+		k_yield();
+	}
+
+	/* Trigger a measurement */
+	SET_BIT(dts->CFGR1, DTS_CFGR1_TS1_START);
+	CLEAR_BIT(dts->CFGR1, DTS_CFGR1_TS1_START);
+
+	/* Wait for interrupt */
+	k_sem_take(&data->sem_isr, K_FOREVER);
```

Comment:
Non-blocker suggestion: should we have a fail-safe on the `while-loop` and `k_sem_take()`? I'm thinking of returning `-ETIMEDOUT` or similar in case the happy-path does not work.

Reply:
Do we really need to do that with `K_FOREVER`. Looking at the documentation, it could return:
- `0` – Semaphore taken.
- `-EBUSY` – Returned without waiting.
- `-EAGAIN` – Waiting period timed out, or the semaphore was reset during the waiting period.

It seems that the `-EBUSY` case can only happens when using `K_NO_WAIT`. For `-EAGAIN`, the period timeout can't happen as `K_FOREVER` is used, and the semaphore reset also can't happened as it is only initialized in `stm32_digi_temp_init` as part of the driver initialization.

Therefore I don't think we need to add a while loop around it. In addition, I haven't found such a structure (with `K_FOREVER`)  elsewhere in the Zephyr's code.

Reply:
Perhaps I didn't explain myself: 

WRT the while-loop: I'm referring to
``` c
	while (READ_BIT(dts->SR, DTS_SR_TS1_RDY) == 0) {
		k_yield();
	}
```

WRT the semaphore: I'm alluding of a scenario where the semaphore is not given, therefore the execution path is blockde. Handling such case involves changing the semaphore from K_FOREVER to a specific timeout and handling the error-codes you described.